File: sdk/typescript/src/namespaces/pipeline.ts
/**
 * Pipeline Namespace API (Idea-to-Execution)
 *
 * Provides endpoints for running and monitoring the 4-stage
 * idea-to-execution pipeline: Ideas → Goals → Workflows → Orchestration.
 */

import type { AragoraClient } from '../client';

/** Pipeline run configuration */
export interface PipelineRunRequest {
  input_text: string;
  stages?: string[];
  debate_rounds?: number;
  workflow_mode?: 'quick' | 'debate';
  dry_run?: boolean;
  enable_receipts?: boolean;
}

/** Pipeline run response with initial status */
export interface PipelineRunResponse {
  pipeline_id: string;
  status: string;
  stages: string[];
}

/** Per-stage status within a pipeline */
export interface PipelineStageStatus {
  stage: string;
  status: string;
  duration?: number;
  error?: string;
}

/** Pipeline status response */
export interface PipelineStatusResponse {
  pipeline_id: string;
  overall_status: string;
  stages: PipelineStageStatus[];
}

/** React Flow graph data */
export interface PipelineGraphResponse {
  nodes: Record<string, unknown>[];
  edges: Record<string, unknown>[];
  pipeline_id: string;
  stage?: string;
}

/** Decision receipt for a completed pipeline */
export interface PipelineReceiptResponse {
  pipeline_id: string;
  receipt: Record<string, unknown>;
}

/**
 * Pipeline namespace for idea-to-execution orchestration.
 *
 * @example
 * ```typescript
 * // Start a pipeline run
 * const { pipeline_id } = await client.pipeline.run({ input_text: 'Build a rate limiter' });
 *
 * // Check status
 * const status = await client.pipeline.status(pipeline_id);
 *
 * // Get React Flow graph
 * const graph = await client.pipeline.graph(pipeline_id);
 *
 * // Get completion receipt
 * const receipt = await client.pipeline.receipt(pipeline_id);
 * ```
 */
export class PipelineNamespace {
  constructor(private client: AragoraClient) {}

  /**
   * Start an async pipeline execution.
   *
   * Runs the 4-stage pipeline: ideation → goals → workflow → orchestration.
   */
  async run(request: PipelineRunRequest): Promise<PipelineRunResponse> {
    return this.client.request<PipelineRunResponse>(
      'POST',
      '/api/v1/canvas/pipeline/run',
      { body: request }
    );
  }

  /**
   * Run full pipeline from an ArgumentCartographer debate export.
   */
  async fromDebate(
    cartographerData: Record<string, unknown>,
    autoAdvance: boolean = true,
  ): Promise<Record<string, unknown>> {
    return this.client.request<Record<string, unknown>>(
      'POST',
      '/api/v1/canvas/pipeline/from-debate',
      {
        body: {
          cartographer_data: cartographerData,
          auto_advance: autoAdvance,
        },
      }
    );
  }

  /**
   * Run full pipeline from raw idea strings.
   */
  async fromIdeas(
    ideas: string[],
    autoAdvance: boolean = true,
  ): Promise<Record<string, unknown>> {
    return this.client.request<Record<string, unknown>>(
      'POST',
      '/api/v1/canvas/pipeline/from-ideas',
      {
        body: {
          ideas,
          auto_advance: autoAdvance,
        },
      }
    );
  }

  /**
   * Get pipeline per-stage status.
   */
  async status(pipelineId: string): Promise<PipelineStatusResponse> {
    return this.client.request<PipelineStatusResponse>(
      'GET',
      `/api/v1/canvas/pipeline/${encodeURIComponent(pipelineId)}/status`
    );
  }

  /**
   * Get full pipeline result.
   */
  async get(pipelineId: string): Promise<Record<string, unknown>> {
    return this.client.request<Record<string, unknown>>(
      'GET',
      `/api/v1/canvas/pipeline/${encodeURIComponent(pipelineId)}`
    );
  }

  /**
   * Get React Flow JSON graph for pipeline stages.
   */
  async graph(
    pipelineId: string,
    options?: { stage?: string },
  ): Promise<PipelineGraphResponse> {
    return this.client.request<PipelineGraphResponse>(
      'GET',
      `/api/v1/canvas/pipeline/${encodeURIComponent(pipelineId)}/graph`,
      { params: options }
    );
  }

  /**
   * Get DecisionReceipt for a completed pipeline.
   */
  async receipt(pipelineId: string): Promise<PipelineReceiptResponse> {
    return this.client.request<PipelineReceiptResponse>(
      'GET',
      `/api/v1/canvas/pipeline/${encodeURIComponent(pipelineId)}/receipt`
    );
  }

  /**
   * Advance a pipeline to the next stage.
   */
  async advance(
    pipelineId: string,
    targetStage: string,
  ): Promise<Record<string, unknown>> {
    return this.client.request<Record<string, unknown>>(
      'POST',
      '/api/v1/canvas/pipeline/advance',
      {
        body: {
          pipeline_id: pipelineId,
          target_stage: targetStage,
        },
      }
    );
  }
}

File: aragora/server/openapi/endpoints/pipeline.py
"""Pipeline endpoint definitions for the Idea-to-Execution canvas pipeline."""

from aragora.server.openapi.helpers import STANDARD_ERRORS, AUTH_REQUIREMENTS

_ID_PARAM = {
    "name": "id",
    "in": "path",
    "required": True,
    "schema": {"type": "string"},
    "description": "Pipeline ID",
}

_STAGE_PARAM = {
    "name": "stage",
    "in": "path",
    "required": True,
    "schema": {"type": "string", "enum": ["ideas", "goals", "actions", "orchestration"]},
    "description": "Pipeline stage name",
}

_GRAPH_BODY = {
    "required": True,
    "content": {
        "application/json": {
            "schema": {
                "type": "object",
                "properties": {
                    "nodes": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "id": {"type": "string"},
                                "type": {"type": "string"},
                                "summary": {"type": "string"},
                                "content": {"type": "string"},
                            },
                        },
                    },
                    "edges": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "source_id": {"type": "string"},
                                "target_id": {"type": "string"},
                                "relation": {"type": "string"},
                            },
                        },
                    },
                },
            }
        }
    },
}


def _json_response(description: str) -> dict:
    """Inline 200 response with generic JSON object body."""
    return {
        "description": description,
        "content": {"application/json": {"schema": {"type": "object"}}},
    }


PIPELINE_ENDPOINTS = {
    "/api/v1/canvas/pipeline/from-ideas": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Create pipeline from ideas",
            "operationId": "createPipelineFromIdeas",
            "description": "Create a full 4-stage pipeline from a list of raw idea strings.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": {
                "required": True,
                "content": {
                    "application/json": {
                        "schema": {
                            "type": "object",
                            "required": ["ideas"],
                            "properties": {
                                "ideas": {
                                    "type": "array",
                                    "items": {"type": "string"},
                                    "description": "List of idea strings",
                                }
                            },
                        }
                    }
                },
            },
            "responses": {
                "200": _json_response("Pipeline result"),
                "400": STANDARD_ERRORS["400"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/pipeline/from-debate": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Create pipeline from debate",
            "operationId": "createPipelineFromDebate",
            "description": "Create a pipeline from an ArgumentCartographer debate graph (nodes + edges).",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": _GRAPH_BODY,
            "responses": {
                "200": _json_response("Pipeline result"),
                "400": STANDARD_ERRORS["400"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/pipeline/from-template": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Create pipeline from template",
            "operationId": "createPipelineFromTemplate",
            "description": "Create a pipeline from a named template.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": {
                "required": True,
                "content": {
                    "application/json": {
                        "schema": {
                            "type": "object",
                            "required": ["template_id"],
                            "properties": {
                                "template_id": {"type": "string"},
                                "parameters": {"type": "object"},
                            },
                        }
                    }
                },
            },
            "responses": {
                "200": _json_response("Pipeline result"),
                "400": STANDARD_ERRORS["400"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/pipeline/advance": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Advance pipeline stage",
            "operationId": "advancePipelineStage",
            "description": "Advance a pipeline to the next stage (e.g. ideas -> goals).",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": {
                "required": True,
                "content": {
                    "application/json": {
                        "schema": {
                            "type": "object",
                            "required": ["pipeline_id", "target_stage"],
                            "properties": {
                                "pipeline_id": {"type": "string"},
                                "target_stage": {
                                    "type": "string",
                                    "enum": ["ideas", "goals", "actions", "orchestration"],
                                },
                            },
                        }
                    }
                },
            },
            "responses": {
                "200": _json_response("Advance result"),
                "400": STANDARD_ERRORS["400"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/pipeline/run": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Run async pipeline",
            "operationId": "runPipeline",
            "description": "Start an asynchronous pipeline execution from ideas through all stages.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": {
                "required": True,
                "content": {
                    "application/json": {
                        "schema": {
                            "type": "object",
                            "required": ["ideas"],
                            "properties": {
                                "ideas": {
                                    "type": "array",
                                    "items": {"type": "string"},
                                },
                            },
                        }
                    }
                },
            },
            "responses": {
                "202": {
                    "description": "Pipeline started",
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "object",
                                "properties": {
                                    "pipeline_id": {"type": "string"},
                                    "status": {"type": "string"},
                                },
                            }
                        }
                    },
                },
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/pipeline/extract-goals": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Extract goals from ideas",
            "operationId": "extractGoals",
            "description": "Use AI to extract structured goals from an ideas canvas.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": _GRAPH_BODY,
            "responses": {
                "200": _json_response("Extracted goals"),
                "400": STANDARD_ERRORS["400"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/convert/debate": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Convert debate to ideas canvas",
            "operationId": "convertDebateToCanvas",
            "description": "Convert a debate graph into an ideas-stage canvas.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": _GRAPH_BODY,
            "responses": {
                "200": _json_response("Ideas canvas"),
                "400": STANDARD_ERRORS["400"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/convert/workflow": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Convert workflow to actions canvas",
            "operationId": "convertWorkflowToCanvas",
            "description": "Convert a workflow definition into an actions-stage canvas.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "requestBody": {
                "required": True,
                "content": {
                    "application/json": {
                        "schema": {"type": "object"},
                    }
                },
            },
            "responses": {
                "200": _json_response("Actions canvas"),
                "400": STANDARD_ERRORS["400"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/pipeline/templates": {
        "get": {
            "tags": ["Pipeline"],
            "summary": "List pipeline templates",
            "operationId": "listPipelineTemplates",
            "description": "List available pipeline templates.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "responses": {
                "200": _json_response("Template list"),
            },
        },
    },
    "/api/v1/canvas/pipeline/{id}": {
        "get": {
            "tags": ["Pipeline"],
            "summary": "Get pipeline",
            "operationId": "getPipeline",
            "description": "Get a pipeline result by ID.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "parameters": [_ID_PARAM],
            "responses": {
                "200": _json_response("Pipeline result"),
                "404": STANDARD_ERRORS["404"],
            },
        },
        "put": {
            "tags": ["Pipeline"],
            "summary": "Save pipeline canvas state",
            "operationId": "savePipeline",
            "description": "Save the current canvas state for a pipeline.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "parameters": [_ID_PARAM],
            "requestBody": {
                "required": True,
                "content": {
                    "application/json": {"schema": {"type": "object"}},
                },
            },
            "responses": {
                "200": _json_response("Saved"),
                "404": STANDARD_ERRORS["404"],
                "500": STANDARD_ERRORS["500"],
            },
        },
    },
    "/api/v1/canvas/pipeline/{id}/status": {
        "get": {
            "tags": ["Pipeline"],
            "summary": "Get pipeline stage status",
            "operationId": "getPipelineStatus",
            "description": "Get per-stage completion status for a pipeline.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "parameters": [_ID_PARAM],
            "responses": {
                "200": _json_response("Stage status"),
                "404": STANDARD_ERRORS["404"],
            },
        },
    },
    "/api/v1/canvas/pipeline/{id}/stage/{stage}": {
        "get": {
            "tags": ["Pipeline"],
            "summary": "Get pipeline stage canvas",
            "operationId": "getPipelineStage",
            "description": "Get the canvas data for a specific pipeline stage.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "parameters": [_ID_PARAM, _STAGE_PARAM],
            "responses": {
                "200": _json_response("Stage canvas"),
                "404": STANDARD_ERRORS["404"],
            },
        },
    },
    "/api/v1/canvas/pipeline/{id}/graph": {
        "get": {
            "tags": ["Pipeline"],
            "summary": "Get pipeline React Flow graph",
            "operationId": "getPipelineGraph",
            "description": "Get the React Flow compatible graph JSON for a pipeline.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "parameters": [
                _ID_PARAM,
                {
                    "name": "stage",
                    "in": "query",
                    "schema": {"type": "string"},
                    "description": "Filter to a specific stage",
                },
            ],
            "responses": {
                "200": _json_response("React Flow graph"),
                "404": STANDARD_ERRORS["404"],
            },
        },
    },
    "/api/v1/canvas/pipeline/{id}/receipt": {
        "get": {
            "tags": ["Pipeline"],
            "summary": "Get pipeline decision receipt",
            "operationId": "getPipelineReceipt",
            "description": "Get the cryptographic decision receipt for a completed pipeline.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "parameters": [_ID_PARAM],
            "responses": {
                "200": _json_response("Decision receipt"),
                "404": STANDARD_ERRORS["404"],
            },
        },
    },
    "/api/v1/canvas/pipeline/{id}/approve-transition": {
        "post": {
            "tags": ["Pipeline"],
            "summary": "Approve stage transition",
            "operationId": "approvePipelineTransition",
            "description": "Approve or reject a pending stage transition.",
            "security": AUTH_REQUIREMENTS["optional"]["security"],
            "parameters": [_ID_PARAM],
            "requestBody": {
                "required": True,
                "content": {
                    "application/json": {
                        "schema": {
                            "type": "object",
                            "required": ["transition_id", "approved"],
                            "properties": {
                                "transition_id": {"type": "string"},
                                "approved": {"type": "boolean"},
                                "reason": {"type": "string"},
                            },
                        }
                    }
                },
            },
            "responses": {
                "200": _json_response("Transition result"),
                "404": STANDARD_ERRORS["404"],
                "400": STANDARD_ERRORS["400"],
            },
        },
    },
}
